lib/core/test/dict_test.cpp

```diff
@@ -0,0 +1,557 @@
+}
+
+TEST(Dict, key_iterator_throws_if_element_erased_middle_after_increment) {
+  DimDict dict{{Dim::Y, -4122}, {Dim::Row, 5619}, {Dim::Event, 829}};
+  auto it = dict.keys_begin();
+  ++it;
+  dict.erase(Dim::Row);
+  EXPECT_THROW_DISCARD(*it, std::runtime_error);
+  EXPECT_THROW_DISCARD(++it, std::runtime_error);
+  EXPECT_THROW_DISCARD(it == dict.keys_end(), std::runtime_error);
+}
+
+TEST(Dict, key_iterator_throws_if_element_erased_back_after_increment) {
+  DimDict dict{{Dim::Y, -4122}, {Dim::Row, 5619}, {Dim::Event, 829}};
+  auto it = dict.keys_begin();
+  ++it;
+  dict.erase(Dim::Event);
+  EXPECT_THROW_DISCARD(*it, std::runtime_error);
+  EXPECT_THROW_DISCARD(++it, std::runtime_error);
+  EXPECT_THROW_DISCARD(it == dict.keys_end(), std::runtime_error);
+}
+
+TEST(Dict, key_iterator_does_not_throw_if_created_after_modification) {
+  DimDict dict{{Dim::Event, 94196}};
+  dict.insert_or_assign(Dim::X, -8164);
+  EXPECT_EQ(dict.begin()->second, 94196);
+  dict.erase(Dim::Event);
+  EXPECT_EQ(dict.begin()->second, -8164);
+}
+
+TEST(Dict, value_iterator_produces_correct_values) {
+  DimDict dict{{Dim::Time, 61892}, {Dim::Event, 619}};
+  auto it = dict.values_begin();
+  EXPECT_EQ(*it, 61892);
+  ++it;
+  EXPECT_EQ(*it, 619);
+  ++it;
+  EXPECT_EQ(it, dict.values_end());
+}
+
+TEST(Dict, const_value_iterator_produces_correct_values) {
+  DimDict dict{{Dim::Time, 4561}, {Dim::Event, 76}};
+  // NOLINTNEXTLINE(performance-unnecessary-copy-initialization)
+  const DimDict const_dict(dict);
+  auto it = const_dict.values_begin();
+  EXPECT_EQ(*it, 4561);
+  ++it;
+  EXPECT_EQ(*it, 76);
+  ++it;
+  EXPECT_EQ(it, const_dict.values_end());
+}
+
+TEST(Dict, value_iterator_can_change_values) {
+  DimDict dict{{Dim::Y, -816}, {Dim::Z, -41}};
+  auto it = dict.values_begin();
+  *it = 923;
+  *(++it) = -5289;
+  EXPECT_EQ(dict[Dim::Y], 923);
+  EXPECT_EQ(dict[Dim::Z], -5289);
+}
+
+TEST(Dict, iterator_of_empty_dict_is_end) {
+  DimDict dict;
+  EXPECT_EQ(dict.begin(), dict.end());
+}
+
+TEST(Dict, iterator_produces_correct_keys_and_values) {
+  DimDict dict{{Dim::Time, 61892}, {Dim::Event, 619}};
+  auto it = dict.begin();
+  EXPECT_EQ((*it).first, Dim::Time);
+  EXPECT_EQ((*it).second, 61892);
+  ++it;
+  EXPECT_EQ((*it).first, Dim::Event);
+  EXPECT_EQ((*it).second, 619);
+  ++it;
+  EXPECT_EQ(it, dict.end());
+}
+
+TEST(Dict, iterator_produces_correct_keys_and_values_via_arrow) {
+  DimDict dict{{Dim::Time, 61892}, {Dim::Event, 619}};
+  auto it = dict.begin();
+  EXPECT_EQ(it->first, Dim::Time);
+  EXPECT_EQ(it->second, 61892);
+  ++it;
+  EXPECT_EQ(it->first, Dim::Event);
+  EXPECT_EQ(it->second, 619);
+  ++it;
+  EXPECT_EQ(it, dict.end());
+}
+
+TEST(Dict, iterator_can_change_values) {
+  DimDict dict{{Dim::Position, -51}, {Dim::Row, 827}};
+  auto it = dict.begin();
+  (*it).second = 991;       // cppcheck-suppress unreadVariable
+  (*(++it)).second = -9761; // cppcheck-suppress unreadVariable
```

Comment:
This cppcheck warning is annoying and reoccuring (had similar issue with assigning to view element recently). Why can it not understand  what iterators are? Having to suppress this everywhere is not ideal.

Reply:
This is probably difficult to do. It needs to understand the relationship between the iterator, the dict, and the temporary pair of references.